File: src/sync/entryClient.ts
import type { Hex, SdkEvent } from '../types';
import { SdkError } from '../errors';
import { isHexStrict } from '../utils/hex';
import { errorToDebug, nonOkResponseDetail } from '../utils/httpDebug';

export interface EntryMemo {
  commitment: Hex;
  memo: Hex;
  cid: number | null;
  is_transparent?: boolean;
  asset_id?: Hex | null;
  amount?: Hex | null;
  partial_hash?: Hex | null;
  txhash?: Hex | null;
  created_at?: number | null;
}

export interface EntryNullifier {
  nullifier: Hex;
  created_at?: number | null;
}

interface EntryListResponse<T> {
  code?: number;
  message?: string;
  data?: { data?: T[]; total?: number } | { data?: T[]; total?: number; ready?: boolean };
}

import { joinUrl } from '../utils/url';

/**
 * Append query parameters to a base URL.
 */
const withQuery = (url: string, params: Record<string, string | number | undefined>) => {
  const search = new URLSearchParams();
  for (const [k, v] of Object.entries(params)) {
    if (v == null) continue;
    search.set(k, String(v));
  }
  const qs = search.toString();
  return qs ? `${url}?${qs}` : url;
};

/**
 * Normalize optional hex values (returns null on invalid hex).
 */
const normalizeOptionalHex = (value: unknown): Hex | null | undefined => {
  if (value == null) return undefined;
  if (isHexStrict(value)) return value;
  return null;
};

/**
 * Normalize total counts from API responses.
 */
const normalizeTotal = (value: unknown): number => {
  const n = typeof value === 'number' ? value : typeof value === 'string' ? Number(value) : Number.NaN;
  if (!Number.isFinite(n) || n < 0) return 0;
  return Math.floor(n);
};

/**
 * Validate and normalize a memo row from EntryService.
 */
const normalizeMemoEntry = (raw: any): EntryMemo => {
  if (!raw || typeof raw !== 'object') {
    throw new SdkError('SYNC', 'Invalid entry memo item', { item: raw });
  }
  if (!isHexStrict(raw.commitment)) throw new SdkError('SYNC', 'Invalid entry memo commitment', { commitment: raw.commitment });
  if (!isHexStrict(raw.memo)) throw new SdkError('SYNC', 'Invalid entry memo payload', { memo: raw.memo });
  const cid = raw.cid;
  if (cid != null && !(typeof cid === 'number' && Number.isInteger(cid) && cid >= 0)) {
    throw new SdkError('SYNC', 'Invalid entry memo cid', { cid });
  }
  const createdAt = raw.created_at;
  if (createdAt != null && !(typeof createdAt === 'number' && Number.isInteger(createdAt) && createdAt >= 0)) {
    throw new SdkError('SYNC', 'Invalid entry memo created_at', { created_at: createdAt });
  }
  const isTransparent = raw.is_transparent;
  if (isTransparent != null && typeof isTransparent !== 'boolean') {
    throw new SdkError('SYNC', 'Invalid entry memo is_transparent', { is_transparent: isTransparent });
  }
  const assetId = normalizeOptionalHex(raw.asset_id);
  const amount = normalizeOptionalHex(raw.amount);
  const partialHash = normalizeOptionalHex(raw.partial_hash);
  const txHash = normalizeOptionalHex(raw.txhash);
  return {
    commitment: raw.commitment,
    memo: raw.memo,
    cid: cid ?? null,
    is_transparent: isTransparent ?? undefined,
    asset_id: assetId ?? undefined,
    amount: amount ?? undefined,
    partial_hash: partialHash ?? undefined,
    txhash: txHash ?? undefined,
    created_at: createdAt ?? null,
  };
};

/**
 * Validate and normalize a nullifier row from EntryService.
 */
const normalizeNullifierEntry = (raw: any): EntryNullifier => {
  if (!raw || typeof raw !== 'object') {
    throw new SdkError('SYNC', 'Invalid entry nullifier item', { item: raw });
  }
  if (!isHexStrict(raw.nullifier)) throw new SdkError('SYNC', 'Invalid entry nullifier', { nullifier: raw.nullifier });
  const createdAt = raw.created_at;
  if (createdAt != null && !(typeof createdAt === 'number' && Number.isInteger(createdAt) && createdAt >= 0)) {
    throw new SdkError('SYNC', 'Invalid entry nullifier created_at', { created_at: createdAt });
  }
  return { nullifier: raw.nullifier, created_at: createdAt ?? null };
};

/**
 * Unwrap list payloads and apply structural validation.
 */
const unwrapList = <T>(payload: EntryListResponse<T>, detail: Record<string, unknown>) => {
  if (typeof payload?.code === 'number' && payload.code !== 0) {
    throw new SdkError('SYNC', payload.message || 'EntryService request failed', payload);
  }
  const itemsRaw = payload?.data?.data;
  const totalRaw = payload?.data?.total;
  if (itemsRaw != null && !Array.isArray(itemsRaw)) {
    throw new SdkError('SYNC', 'Invalid entry response: data.data must be an array', { ...detail, data: payload?.data });
  }
  return { items: (itemsRaw ?? []) as T[], total: normalizeTotal(totalRaw) };
};

/**
 * Unwrap list payloads and capture the "ready" flag if present.
 */
const unwrapListWithReady = <T>(payload: EntryListResponse<T>, detail: Record<string, unknown>) => {
  const base = unwrapList(payload, detail);
  const ready = (payload as any)?.data?.ready;
  return { ...base, ready: ready == null ? true : Boolean(ready) };
};

type DebugEmitter = (event: Extract<SdkEvent, { type: 'debug' }>) => void;

/**
 * HTTP client for EntryService memo/nullifier endpoints.
 */
export class EntryClient {
  constructor(
    private readonly baseUrl: string,
    private readonly debugEmit?: DebugEmitter,
  ) {}

  /**
   * Fetch memo pages for a viewing address.
   */
  async listMemos(input: { chainId: number; address: string; offset: number; limit: number; signal?: AbortSignal }) {
    const url = withQuery(joinUrl(this.baseUrl, '/api/v1/viewing/memos/list'), {
      offset: input.offset,
      limit: input.limit,
      chain_id: input.chainId,
      address: input.address,
      order: 'asc',
    });
    this.debugEmit?.({
      type: 'debug',
      payload: { scope: 'http:entry', message: 'request', detail: { method: 'GET', url } },
    });
    let response: Response;
    try {
      response = await fetch(url, { signal: input.signal });
    } catch (error) {
      this.debugEmit?.({ type: 'debug', payload: { scope: 'http:entry', message: 'network_error', detail: { url, error: errorToDebug(error) } } });
      throw error;
    }
    this.debugEmit?.({
      type: 'debug',
      payload: { scope: 'http:entry', message: 'response', detail: { url, status: response.status, ok: response.ok } },
    });
    if (!response.ok) {
      throw new SdkError('SYNC', 'EntryService memos request failed', await nonOkResponseDetail(response, url));
    }
    const payload = (await response.json()) as EntryListResponse<EntryMemo>;
    const { items, total } = unwrapList(payload, { url });
    return { items: items.map(normalizeMemoEntry), total };
  }

  /**
   * Fetch nullifier pages for a viewing address.
   */
  async listNullifiers(input: { chainId: number; address: string; offset: number; limit: number; signal?: AbortSignal }) {
    const url = withQuery(joinUrl(this.baseUrl, '/api/v1/viewing/nullifier/list'), {
      offset: input.offset,
      limit: input.limit,
      chain_id: input.chainId,
      address: input.address,
      order: 'asc',
    });
    this.debugEmit?.({
      type: 'debug',
      payload: { scope: 'http:entry', message: 'request', detail: { method: 'GET', url } },
    });
    let response: Response;
    try {
      response = await fetch(url, { signal: input.signal });
    } catch (error) {
      this.debugEmit?.({ type: 'debug', payload: { scope: 'http:entry', message: 'network_error', detail: { url, error: errorToDebug(error) } } });
      throw error;
    }
    this.debugEmit?.({
      type: 'debug',
      payload: { scope: 'http:entry', message: 'response', detail: { url, status: response.status, ok: response.ok } },
    });
    if (!response.ok) {
      throw new SdkError('SYNC', 'EntryService nullifier request failed', await nonOkResponseDetail(response, url));
    }
    const payload = (await response.json()) as EntryListResponse<EntryNullifier>;
    const { items, total } = unwrapList(payload, { url });
    return { items: items.map(normalizeNullifierEntry), total };
  }

  /**
   * Fetch nullifiers with the block-indexed pagination API.
   * This endpoint may return a "ready" flag indicating if more pages are expected.
   */
  async listNullifiersByBlock(input: { chainId: number; address: string; offset: number; limit: number; signal?: AbortSignal }) {
    const url = withQuery(joinUrl(this.baseUrl, '/api/v1/viewing/nullifier/list_by_block'), {
      offset: input.offset,
      limit: input.limit,
      chain_id: input.chainId,
      address: input.address,
      order: 'asc',
    });
    this.debugEmit?.({
      type: 'debug',
      payload: { scope: 'http:entry', message: 'request', detail: { method: 'GET', url } },
    });
    let response: Response;
    try {
      response = await fetch(url, { signal: input.signal });
    } catch (error) {
      this.debugEmit?.({ type: 'debug', payload: { scope: 'http:entry', message: 'network_error', detail: { url, error: errorToDebug(error) } } });
      throw error;
    }
    this.debugEmit?.({
      type: 'debug',
      payload: { scope: 'http:entry', message: 'response', detail: { url, status: response.status, ok: response.ok } },
    });
    if (!response.ok) {
      throw new SdkError('SYNC', 'EntryService nullifier list_by_block request failed', await nonOkResponseDetail(response, url));
    }
    const payload = (await response.json()) as EntryListResponse<EntryNullifier>;
    const { items, total, ready } = unwrapListWithReady(payload, { url });
    return { items: items.map(normalizeNullifierEntry), total, ready };
  }
}

File: core/crates/ocash-sync/src/entry_client.rs
//! HTTP client for the OCash Entry service.
//!
//! Endpoints:
//! - GET /api/v1/viewing/memos/list
//! - GET /api/v1/viewing/nullifier/list_by_block

use serde::{Deserialize, Serialize};
use ocash_types::{Hex, OcashError, Result};
use std::time::Duration;

/// An entry memo from the Entry service.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct EntryMemo {
    pub cid: u64,
    pub commitment: Hex,
    pub memo: Hex,
}

/// An entry nullifier from the Entry service.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct EntryNullifier {
    pub nid: u64,
    pub nullifier: Hex,
}

/// Entry service API response wrapper.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct EntryResponse<T> {
    pub code: Option<i32>,
    pub message: Option<String>,
    pub data: EntryResponseData<T>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct EntryResponseData<T> {
    pub data: Vec<T>,
    pub total: u64,
    pub ready: Option<bool>,
}

/// List memos query result.
pub struct ListMemosResult {
    pub items: Vec<EntryMemo>,
    pub total: u64,
}

/// List nullifiers query result.
pub struct ListNullifiersResult {
    pub items: Vec<EntryNullifier>,
    pub total: u64,
    pub ready: bool,
}

/// Entry service client.
pub struct EntryClient {
    base_url: String,
    client: reqwest::Client,
    timeout: Duration,
}

impl EntryClient {
    pub fn new(base_url: &str, timeout_ms: Option<u64>) -> Self {
        let timeout_ms = timeout_ms.unwrap_or(20_000);
        Self {
            base_url: base_url.trim_end_matches('/').to_string(),
            client: reqwest::Client::builder()
                .timeout(Duration::from_millis(timeout_ms))
                .build()
                .unwrap_or_default(),
            timeout: Duration::from_millis(timeout_ms),
        }
    }

    /// Fetch memos from the Entry service.
    ///
    /// GET /api/v1/viewing/memos/list?chain_id=X&address=ADDR&offset=N&limit=L&order=asc
    pub async fn list_memos(
        &self,
        chain_id: u64,
        contract_address: &str,
        offset: u64,
        limit: u64,
    ) -> Result<ListMemosResult> {
        let url = format!(
            "{}/api/v1/viewing/memos/list?chain_id={}&address={}&offset={}&limit={}&order=asc",
            self.base_url, chain_id, contract_address, offset, limit
        );

        let resp = self.client
            .get(&url)
            .timeout(self.timeout)
            .send()
            .await
            .map_err(|e| OcashError::Other(format!("entry request failed: {}", e)))?;

        if !resp.status().is_success() {
            return Err(OcashError::Other(format!(
                "entry service returned status {}",
                resp.status()
            )));
        }

        let body: EntryResponse<EntryMemo> = resp
            .json()
            .await
            .map_err(|e| OcashError::Other(format!("failed to parse entry response: {}", e)))?;

        Ok(ListMemosResult {
            items: body.data.data,
            total: body.data.total,
        })
    }

    /// Fetch nullifiers from the Entry service.
    ///
    /// GET /api/v1/viewing/nullifier/list_by_block?chain_id=X&address=ADDR&offset=N&limit=L&order=asc
    pub async fn list_nullifiers_by_block(
        &self,
        chain_id: u64,
        contract_address: &str,
        offset: u64,
        limit: u64,
    ) -> Result<ListNullifiersResult> {
        let url = format!(
            "{}/api/v1/viewing/nullifier/list_by_block?chain_id={}&address={}&offset={}&limit={}&order=asc",
            self.base_url, chain_id, contract_address, offset, limit
        );

        let resp = self.client
            .get(&url)
            .timeout(self.timeout)
            .send()
            .await
            .map_err(|e| OcashError::Other(format!("entry request failed: {}", e)))?;

        if !resp.status().is_success() {
            return Err(OcashError::Other(format!(
                "entry service returned status {}",
                resp.status()
            )));
        }

        let body: EntryResponse<EntryNullifier> = resp
            .json()
            .await
            .map_err(|e| OcashError::Other(format!("failed to parse entry response: {}", e)))?;

        Ok(ListNullifiersResult {
            items: body.data.data,
            total: body.data.total,
            ready: body.data.ready.unwrap_or(true),
        })
    }
}
